### DISCUSSION TITLE
11 hour m4b has taken over  an hour

### DISCUSSION THREAD

**Discussion by gordonrjones** [enzo1982/freac#446] [Help & Support] [answered]
11 hour m4b has taken over  an hour so far and still has that long to finish, it is the first time it has ever taken this long, has anyone ever encountered this problem?

**Comment by AlanHK** [comment]
If you're doing any signal processing, and making an M4B, you have an option under "Conversion to single output file":
"Process each track individually" or
"Treat the combined track as a single track".

Individual is much faster, some analysis over an entire book length file takes a very long time.

**Comment by enzo1982** [comment, marked as answer]
@AlanHK, there should be no difference in processing speed between the two options actually. The only difference is that with the _individual_ option, DSP components are restarted for each track, while with the _combined_ option, all tracks are fed to the same components serially.

The _combined_ option should be used for gapless albums to avoid discontinuities at track boundaries. The _individual_ option can be used to ensure that DSP filter state from one track does not influence the next track in any way. In most cases this does not make a real difference, though, so the _combined_ option is recommended (and it's also the default).

@gordonrjones, a conversion of an 11 hour m4b taking that long is definitely not normal. Did it ever finish and was the resulting file ok? If it is possible for you to make the source m4b file available to me (e.g. via Dropbox or Google Drive), please do so, so I can try and find out what is causing this. You can send a link to the file to support@freac.org.

**Reply by enzo1982** [reply]
@gordonrjones, thank you for providing the files!

I can confirm that converting the large M4B file takes unusually long here too. The estimated conversion time starts out pretty low and then goes up over time. This happens independent of the AAC decoder used (FDK-AAC or FAAD2) and the target format of the conversion. Based on that, I would guess that the issue lies in the MP4v2 demuxer library, but further investigation will be needed to pin down the actual issue.

I'll investigate and then report back.

**Reply by gordonrjones** [reply]
Thank you.

**Reply by enzo1982** [reply]
A fix for this is now available in the latest [continuous build](https://github.com/enzo1982/freac/releases/continuous). This will also be included in the upcoming fre:ac 1.1.7.

The slowdown was indeed caused by the MP4v2 library.

Normally, MP4 files consist of many small chunks of samples. For example, the file fre:ac created when you converted the 11 hour .m4b consists of 42.500 individual chunks with 47 samples each.

The source file, however, contains only a single chunk with 1.8 million samples. This causes MP4v2's algorithm for getting the position of a sample in the file to slow down extremely. The algorithm first finds the position of the chunk containing the requested sample and then adds the sizes of all samples in the chunk up to that sample. This works reasonably well when each chunk has only a small number of samples, but with a huge chunk, it will add all the sample sizes starting with the first sample over and over.

The fix consists of a cache that stores the previously requested sample number and position. When the next sample is requested, only the size of the previous sample has to be added to the saved position to find the next position.

This speeds up conversion of the 11 hour .m4b file from 1 hour 56 minutes to just 1 minute 30 seconds on my system. When using the channel and sample rate converters (which you apparently did to get 48 kHz Mono output), the conversion is sped up from 2 hours 2 minutes to just 8 minutes.

**Reply by gordonrjones** [reply]
Works perfect now thanks a lot...
